File: slp/backend-dotnet/Features/Quiz/QuizRepository.cs
using backend_dotnet.Data;
using backend_dotnet.Features.Dashboard;
using Microsoft.EntityFrameworkCore;

namespace backend_dotnet.Features.Quiz;

public class QuizRepository : IQuizRepository
{
    private readonly AppDbContext _context;

    public QuizRepository(AppDbContext context)
    {
        _context = context;
    }

    public async Task<Quiz?> GetByIdAsync(int id, bool includeDisabled = false)
    {
        IQueryable<Quiz> query = _context.Quizzes
            .Include(q => q.QuizQuestions.OrderBy(qq => qq.DisplayOrder))
            .Include(q => q.QuizTags).ThenInclude(qt => qt.Tag)
            .Include(q => q.QuizSources).ThenInclude(qs => qs.Source)
            .Include(q => q.QuizNotes).ThenInclude(qn => qn.Note)
            .Include(q => q.User);

        if (includeDisabled)
        {
            // Bypass the global query filter to include disabled quizzes
            query = query.IgnoreQueryFilters();
        }
        else
        {
            // Explicitly exclude disabled quizzes (global filter may also apply)
            query = query.Where(q => !q.Disabled);
        }

        return await query.FirstOrDefaultAsync(q => q.Id == id);
    }

    public async Task<IEnumerable<Quiz>> GetUserQuizzesAsync(int userId, bool includeDisabled = false)
    {
        var query = _context.Quizzes.Where(q => q.UserId == userId);
        if (!includeDisabled)
            query = query.Where(q => !q.Disabled);
        return await query
            .Include(q => q.User)                   // ← ADD THIS
            .Include(q => q.QuizQuestions)          // ← ADD THIS
            .Include(q => q.QuizTags).ThenInclude(qt => qt.Tag)
            .OrderByDescending(q => q.UpdatedAt)
            .ToListAsync();
    }

    public async Task<IEnumerable<Quiz>> GetPublicQuizzesAsync(string? visibility = null, bool includeDisabled = false)
    {
        var query = _context.Quizzes.AsQueryable();
        if (!includeDisabled)
            query = query.Where(q => !q.Disabled);

        if (string.IsNullOrEmpty(visibility) || visibility == "public")
            query = query.Where(q => q.Visibility == "public");
        else if (visibility == "unlisted")
            query = query.Where(q => q.Visibility == "unlisted");

        return await query
            .Include(q => q.User)
            .Include(q => q.QuizQuestions)          // ← ADD THIS
            .Include(q => q.QuizTags).ThenInclude(qt => qt.Tag)
            .OrderByDescending(q => q.CreatedAt)
            .ToListAsync();
    }

    public async Task<Quiz> CreateAsync(Quiz quiz)
    {
        _context.Quizzes.Add(quiz);
        await _context.SaveChangesAsync();
        return quiz;
    }

    public async Task UpdateAsync(Quiz quiz)
    {
        quiz.UpdatedAt = DateTime.UtcNow;
        _context.Quizzes.Update(quiz);
        await _context.SaveChangesAsync();
    }

    // REPLACE the existing SoftDeleteAsync:
    public async Task SoftDeleteAsync(int id)
    {
        var quiz = await _context.Quizzes.FindAsync(id);
        if (quiz != null)
        {
            quiz.Disabled = true;
            quiz.UpdatedAt = DateTime.UtcNow;
            await _context.SaveChangesAsync();
        }
    }

    // ADD a true hard delete — this is what DELETE /api/quiz/{id} should call:
    public async Task HardDeleteAsync(int id)
    {
        // WHY raw SQL instead of _context.Quizzes.Remove():
        //
        // DeleteQuizAsync calls GetByIdAsync before this method, which loads the Quiz
        // entity WITH its QuizQuestions into EF's change tracker. Calling Remove() then
        // causes EF to sever the in-memory FK relationship before issuing the SQL DELETE.
        // Because quiz_id is non-nullable (required relationship), EF throws:
        //   InvalidOperationException: "association has been severed but relationship is required"
        //
        // Bypassing EF tracking with raw SQL avoids this entirely.
        // Postgres ON DELETE CASCADE on quiz_question, quiz_attempt, quiz_attempt_answer,
        // quiz_tag, quiz_note, quiz_source, quiz_view removes all child rows automatically.
        await _context.Database.ExecuteSqlRawAsync(
            "DELETE FROM quiz WHERE id = {0}", id);
    }

    public async Task<bool> ExistsAsync(int id)
    {
        return await _context.Quizzes.AnyAsync(q => q.Id == id && !q.Disabled);
    }

    public async Task<(IEnumerable<Quiz> Items, int TotalCount)> SearchAsync(
        string? searchTerm,
        int? userId,
        string? visibility,
        bool includeDisabled,
        string? sortBy,
        string? sortOrder,
        int page,
        int pageSize)
    {
        var query = _context.Quizzes.AsQueryable();

        // Apply disabled filter correctly
        if (includeDisabled)
        {
            query = query.IgnoreQueryFilters(); // Bypass global filter
        }
        else
        {
            query = query.Where(q => !q.Disabled);
        }

        // Filter by user
        if (userId.HasValue)
            query = query.Where(q => q.UserId == userId.Value);

        // Filter by visibility
        if (!string.IsNullOrEmpty(visibility))
            query = query.Where(q => q.Visibility == visibility);
        else if (!userId.HasValue) // public listing, default to public
            query = query.Where(q => q.Visibility == "public");

        // Search term
        if (!string.IsNullOrWhiteSpace(searchTerm))
        {
            var pattern = $"%{searchTerm}%";
            query = query.Where(q => EF.Functions.ILike(q.Title, pattern) ||
                                     (q.Description != null && EF.Functions.ILike(q.Description, pattern)));
        }

        // Total count (before sorting/includes)
        var totalCount = await query.CountAsync();

        // Sorting
        IOrderedQueryable<Quiz> orderedQuery = (sortBy?.ToLower(), sortOrder?.ToLower()) switch
        {
            ("title", "asc") => query.OrderBy(q => q.Title),
            ("title", "desc") => query.OrderByDescending(q => q.Title),
            _ => sortOrder == "asc" ? query.OrderBy(q => q.CreatedAt) : query.OrderByDescending(q => q.CreatedAt)
        };

        // Apply pagination and includes
        var items = await orderedQuery
            .Include(q => q.User)
            .Include(q => q.QuizQuestions)
            .Include(q => q.QuizTags).ThenInclude(qt => qt.Tag)
            .Skip((page - 1) * pageSize)
            .Take(pageSize)
            .ToListAsync();

        return (items, totalCount);
    }

    // ==================== QuizQuestion repository methods ====================

    public async Task<IEnumerable<QuizQuestion>> GetQuestionsByQuizIdAsync(int quizId)
    {
        return await _context.QuizQuestions
            .Where(qq => qq.QuizId == quizId)
            .OrderBy(qq => qq.DisplayOrder)
            .ToListAsync();
    }

    public async Task<QuizQuestion?> GetQuizQuestionByIdAsync(int id)
    {
        return await _context.QuizQuestions
            .Include(qq => qq.Quiz)
            .FirstOrDefaultAsync(qq => qq.Id == id);
    }

    public async Task<QuizQuestion> CreateQuizQuestionAsync(QuizQuestion quizQuestion)
    {
        quizQuestion.CreatedAt = DateTime.UtcNow;
        quizQuestion.UpdatedAt = DateTime.UtcNow;
        _context.QuizQuestions.Add(quizQuestion);
        await _context.SaveChangesAsync();
        return quizQuestion;
    }

    public async Task UpdateQuizQuestionAsync(QuizQuestion quizQuestion)
    {
        quizQuestion.UpdatedAt = DateTime.UtcNow;
        _context.QuizQuestions.Update(quizQuestion);
        await _context.SaveChangesAsync();
    }

    public async Task DeleteQuizQuestionAsync(int id)
    {
        var question = await _context.QuizQuestions.FindAsync(id);
        if (question != null)
        {
            _context.QuizQuestions.Remove(question);
            await _context.SaveChangesAsync();
        }
    }

    public async Task ReorderQuizQuestionsAsync(int quizId, List<int> questionIds)
    {
        var questions = await _context.QuizQuestions
            .Where(qq => qq.QuizId == quizId)
            .ToListAsync();

        for (int i = 0; i < questionIds.Count; i++)
        {
            var q = questions.FirstOrDefault(qq => qq.Id == questionIds[i]);
            if (q != null)
            {
                q.DisplayOrder = i + 1;
                q.UpdatedAt = DateTime.UtcNow;
            }
        }
        await _context.SaveChangesAsync();
    }

    public async Task<IEnumerable<Note.Note>> GetNotesByQuizIdAsync(int quizId)
    {
        var quiz = await _context.Quizzes
            .Include(q => q.QuizNotes)
                .ThenInclude(qn => qn.Note)
            .FirstOrDefaultAsync(q => q.Id == quizId && !q.Disabled);
        return quiz?.QuizNotes.Select(qn => qn.Note) ?? Enumerable.Empty<Note.Note>();
    }

    public async Task AddNoteToQuizAsync(int quizId, int noteId)
    {
        var quizNote = new QuizNote { QuizId = quizId, NoteId = noteId };
        _context.QuizNotes.Add(quizNote);
        await _context.SaveChangesAsync();
    }

    public async Task RemoveNoteFromQuizAsync(int quizId, int noteId)
    {
        var quizNote = await _context.QuizNotes
            .FirstOrDefaultAsync(qn => qn.QuizId == quizId && qn.NoteId == noteId);
        if (quizNote != null)
        {
            _context.QuizNotes.Remove(quizNote);
            await _context.SaveChangesAsync();
        }
    }

    public async Task<Note.Note> CreateNoteAndAddToQuizAsync(int quizId, int userId, string title, string content)
    {
        var note = new Note.Note
        {
            UserId = userId,
            Title = title,
            Content = content,
            CreatedAt = DateTime.UtcNow,
            UpdatedAt = DateTime.UtcNow
        };
        _context.Notes.Add(note);
        await _context.SaveChangesAsync(); // save to generate Id

        await AddNoteToQuizAsync(quizId, note.Id);
        return note;
    }

    public async Task<Note.Note?> GetNoteByIdAndUserAsync(int noteId, int userId)
    {
        return await _context.Notes
            .FirstOrDefaultAsync(n => n.Id == noteId && n.UserId == userId);
    }

    public async Task<IEnumerable<Source.Source>> GetSourcesByQuizIdAsync(int quizId)
    {
        var quiz = await _context.Quizzes
            .Include(q => q.QuizSources)
                .ThenInclude(qs => qs.Source)
            .FirstOrDefaultAsync(q => q.Id == quizId && !q.Disabled);

        return quiz?.QuizSources.Select(qs => qs.Source) ?? Enumerable.Empty<Source.Source>();
    }

    public async Task AddSourceToQuizAsync(int quizId, int sourceId)
    {
        // Avoid duplicates
        var exists = await _context.QuizSources
            .AnyAsync(qs => qs.QuizId == quizId && qs.SourceId == sourceId);

        if (!exists)
        {
            var quizSource = new QuizSource { QuizId = quizId, SourceId = sourceId };
            _context.QuizSources.Add(quizSource);
            await _context.SaveChangesAsync();
        }
    }

    public async Task RemoveSourceFromQuizAsync(int quizId, int sourceId)
    {
        var quizSource = await _context.QuizSources
            .FirstOrDefaultAsync(qs => qs.QuizId == quizId && qs.SourceId == sourceId);

        if (quizSource != null)
        {
            _context.QuizSources.Remove(quizSource);
            await _context.SaveChangesAsync();
        }
    }

    public async Task<IEnumerable<Quiz>> GetAllForAdminAsync(string? search = null)
    {
        var query = _context.Quizzes
            .IgnoreQueryFilters()
            .Include(q => q.User)
            .AsQueryable();

        if (!string.IsNullOrWhiteSpace(search))
        {
            var pattern = $"%{search}%";
            query = query.Where(q =>
                EF.Functions.ILike(q.Title, pattern) ||
                (q.User != null && EF.Functions.ILike(q.User.Username, pattern)));
        }

        return await query
            .OrderByDescending(q => q.CreatedAt)
            .ToListAsync();
    }

    public async Task<List<TopQuizDto>> GetTopQuizzesByAttemptsAsync(int limit)
    {
        var query = from q in _context.Quizzes
                    where q.Visibility == "public" && !q.Disabled
                    join user in _context.Users on q.UserId equals user.Id into userJoin
                    from user in userJoin.DefaultIfEmpty()
                    select new
                    {
                        Quiz = q,
                        AuthorUsername = user != null ? user.Username : "[deleted]",
                        AttemptCount = _context.QuizAttempts.Count(a => a.QuizId == q.Id && a.Status == "completed"),
                        CommentCount = _context.Comments.Count(c => c.TargetType == "quiz" && c.TargetId == q.Id && c.DeletedAt == null),
                        QuestionCount = _context.QuizQuestions.Count(qq => qq.QuizId == q.Id)
                    };

        var results = await query
            .OrderByDescending(x => x.AttemptCount)
            .Take(limit)
            .Select(x => new TopQuizDto
            {
                Id = x.Quiz.Id,
                Title = x.Quiz.Title,
                AuthorUsername = x.AuthorUsername,
                AttemptCount = x.AttemptCount,
                CommentCount = x.CommentCount,
                QuestionCount = x.QuestionCount
            })
            .ToListAsync();

        return results;
    }
}

File: slp/backend-nestjs/src/modules/quiz/quiz.repository.ts
import { Injectable } from "@nestjs/common";
import { InjectRepository } from "@nestjs/typeorm";
import { Repository, SelectQueryBuilder } from "typeorm";
import { Quiz } from "./quiz.entity";
import { QuizQuestion } from "./quiz-question.entity";
import { QuizTag } from "./quiz-tag.entity";
import { QuizNote } from "./quiz-note.entity";
import { QuizSource } from "./quiz-source.entity";
import { Note } from "../note/note.entity";
import { Source } from "../source/source.entity";
import { Tag } from "../tag/tag.entity";
import { TopQuizDto } from "../dashboard/dto/top-quiz.dto"; // Adjust import path

@Injectable()
export class QuizRepository {
  constructor(
    @InjectRepository(Quiz)
    private quizRepo: Repository<Quiz>,
    @InjectRepository(QuizQuestion)
    private quizQuestionRepo: Repository<QuizQuestion>,
    @InjectRepository(QuizTag)
    private quizTagRepo: Repository<QuizTag>,
    @InjectRepository(QuizNote)
    private quizNoteRepo: Repository<QuizNote>,
    @InjectRepository(QuizSource)
    private quizSourceRepo: Repository<QuizSource>,
    @InjectRepository(Note)
    private noteRepo: Repository<Note>,
    @InjectRepository(Source)
    private sourceRepo: Repository<Source>,
  ) {}

  async getById(id: number, includeDisabled = false): Promise<Quiz | null> {
    let query = this.quizRepo
      .createQueryBuilder("quiz")
      .leftJoinAndSelect("quiz.user", "user")
      .leftJoinAndSelect("quiz.quizQuestions", "quizQuestions")
      .leftJoinAndSelect("quiz.quizTags", "quizTags")
      .leftJoinAndSelect("quizTags.tag", "tag")
      .leftJoinAndSelect("quiz.quizNotes", "quizNotes")
      .leftJoinAndSelect("quizNotes.note", "note")
      .leftJoinAndSelect("quiz.quizSources", "quizSources")
      .leftJoinAndSelect("quizSources.source", "source")
      .where("quiz.id = :id", { id });

    if (!includeDisabled) {
      query = query.andWhere("quiz.disabled = false");
    }

    return query.getOne();
  }

  async getUserQuizzes(
    userId: number,
    includeDisabled = false,
  ): Promise<Quiz[]> {
    let query = this.quizRepo
      .createQueryBuilder("quiz")
      .leftJoinAndSelect("quiz.user", "user")
      .leftJoinAndSelect("quiz.quizQuestions", "quizQuestions")
      .leftJoinAndSelect("quiz.quizTags", "quizTags")
      .leftJoinAndSelect("quizTags.tag", "tag")
      .where("quiz.userId = :userId", { userId });

    if (!includeDisabled) {
      query = query.andWhere("quiz.disabled = false");
    }

    return query.orderBy("quiz.updatedAt", "DESC").getMany();
  }

  async getPublicQuizzes(
    visibility?: string,
    includeDisabled = false,
  ): Promise<Quiz[]> {
    let query = this.quizRepo
      .createQueryBuilder("quiz")
      .leftJoinAndSelect("quiz.user", "user")
      .leftJoinAndSelect("quiz.quizQuestions", "quizQuestions")
      .leftJoinAndSelect("quiz.quizTags", "quizTags")
      .leftJoinAndSelect("quizTags.tag", "tag");

    if (!includeDisabled) {
      query = query.andWhere("quiz.disabled = false");
    }

    if (!visibility || visibility === "public") {
      query = query.andWhere("quiz.visibility = :visibility", {
        visibility: "public",
      });
    } else if (visibility === "unlisted") {
      query = query.andWhere("quiz.visibility = :visibility", {
        visibility: "unlisted",
      });
    }

    return query.orderBy("quiz.createdAt", "DESC").getMany();
  }

  async create(quiz: Quiz): Promise<Quiz> {
    const saved = await this.quizRepo.save(quiz);
    const reloaded = await this.getById(saved.id);

    if (!reloaded) {
      throw new Error(`Failed to reload quiz after creation (ID: ${saved.id})`);
    }

    return reloaded;
  }

  async update(quiz: Quiz): Promise<void> {
    quiz.updatedAt = new Date();
    await this.quizRepo.save(quiz);
  }

  async hardDelete(id: number): Promise<void> {
    await this.quizRepo.delete(id);
  }

  async exists(id: number): Promise<boolean> {
    const count = await this.quizRepo.count({ where: { id, disabled: false } });
    return count > 0;
  }

  async search(
    searchTerm?: string,
    userId?: number,
    visibility?: string,
    includeDisabled = false,
    sortBy?: string,
    sortOrder?: "asc" | "desc",
    page = 1,
    pageSize = 20,
  ): Promise<{ items: Quiz[]; totalCount: number }> {
    let query = this.quizRepo
      .createQueryBuilder("quiz")
      .leftJoinAndSelect("quiz.user", "user")
      .leftJoinAndSelect("quiz.quizQuestions", "quizQuestions")
      .leftJoinAndSelect("quiz.quizTags", "quizTags")
      .leftJoinAndSelect("quizTags.tag", "tag");

    if (!includeDisabled) {
      query = query.andWhere("quiz.disabled = false");
    } else {
      // If includeDisabled, we must also include disabled quizzes
      // (no filter on disabled)
    }

    if (userId) {
      query = query.andWhere("quiz.userId = :userId", { userId });
    }

    if (visibility) {
      query = query.andWhere("quiz.visibility = :visibility", { visibility });
    } else if (!userId) {
      // Public listing, default to public
      query = query.andWhere("quiz.visibility = :visibility", {
        visibility: "public",
      });
    }

    if (searchTerm) {
      query = query.andWhere(
        "(quiz.title ILIKE :search OR quiz.description ILIKE :search)",
        { search: `%${searchTerm}%` },
      );
    }

    // Sorting
    const order = sortOrder === "asc" ? "ASC" : "DESC";
    if (sortBy === "title") {
      query = query.orderBy("quiz.title", order);
    } else {
      query = query.orderBy("quiz.createdAt", order);
    }

    const totalCount = await query.getCount();

    const items = await query
      .skip((page - 1) * pageSize)
      .take(pageSize)
      .getMany();

    return { items, totalCount };
  }

  // ---------- QuizQuestion ----------
  async getQuestionsByQuizId(quizId: number): Promise<QuizQuestion[]> {
    return this.quizQuestionRepo.find({
      where: { quizId },
      order: { displayOrder: "ASC" },
    });
  }

  async getQuizQuestionById(id: number): Promise<QuizQuestion | null> {
    return this.quizQuestionRepo.findOne({
      where: { id },
      relations: ["quiz"],
    });
  }

  async createQuizQuestion(quizQuestion: QuizQuestion): Promise<QuizQuestion> {
    quizQuestion.createdAt = new Date();
    quizQuestion.updatedAt = new Date();
    return this.quizQuestionRepo.save(quizQuestion);
  }

  async updateQuizQuestion(quizQuestion: QuizQuestion): Promise<void> {
    quizQuestion.updatedAt = new Date();
    await this.quizQuestionRepo.save(quizQuestion);
  }

  async deleteQuizQuestion(id: number): Promise<void> {
    await this.quizQuestionRepo.delete(id);
  }

  async reorderQuizQuestions(
    quizId: number,
    questionIds: number[],
  ): Promise<void> {
    const questions = await this.quizQuestionRepo.find({ where: { quizId } });
    for (let i = 0; i < questionIds.length; i++) {
      const q = questions.find((qq) => qq.id === questionIds[i]);
      if (q) {
        q.displayOrder = i + 1;
        q.updatedAt = new Date();
      }
    }
    await this.quizQuestionRepo.save(questions);
  }

  // ---------- Notes ----------
  async getNotesByQuizId(quizId: number): Promise<Note[]> {
    const quiz = await this.quizRepo.findOne({
      where: { id: quizId, disabled: false },
      relations: ["quizNotes", "quizNotes.note"],
    });
    return quiz?.quizNotes.map((qn) => qn.note) ?? [];
  }

  async addNoteToQuiz(quizId: number, noteId: number): Promise<void> {
    const quizNote = this.quizNoteRepo.create({ quizId, noteId });
    await this.quizNoteRepo.save(quizNote);
  }

  async removeNoteFromQuiz(quizId: number, noteId: number): Promise<void> {
    await this.quizNoteRepo.delete({ quizId, noteId });
  }

  async createNoteAndAddToQuiz(
    quizId: number,
    userId: number,
    title: string,
    content: string,
  ): Promise<Note> {
    const note = this.noteRepo.create({
      userId,
      title,
      content,
      createdAt: new Date(),
      updatedAt: new Date(),
    });
    await this.noteRepo.save(note);
    await this.addNoteToQuiz(quizId, note.id);
    return note;
  }

  async getNoteByIdAndUser(
    noteId: number,
    userId: number,
  ): Promise<Note | null> {
    return this.noteRepo.findOne({ where: { id: noteId, userId } });
  }

  // ---------- Sources ----------
  async getSourcesByQuizId(quizId: number): Promise<Source[]> {
    const quiz = await this.quizRepo.findOne({
      where: { id: quizId, disabled: false },
      relations: ["quizSources", "quizSources.source"],
    });
    return quiz?.quizSources.map((qs) => qs.source) ?? [];
  }

  async addSourceToQuiz(quizId: number, sourceId: number): Promise<void> {
    const exists = await this.quizSourceRepo.findOne({
      where: { quizId, sourceId },
    });
    if (!exists) {
      const quizSource = this.quizSourceRepo.create({ quizId, sourceId });
      await this.quizSourceRepo.save(quizSource);
    }
  }

  async removeSourceFromQuiz(quizId: number, sourceId: number): Promise<void> {
    await this.quizSourceRepo.delete({ quizId, sourceId });
  }

  // ---------- Admin ----------
  async getAllForAdminPaginated(
    search?: string,
    page = 1,
    pageSize = 20,
  ): Promise<{ items: Quiz[]; total: number }> {
    const query = this.quizRepo
      .createQueryBuilder("quiz")
      .leftJoinAndSelect("quiz.user", "user");

    if (search) {
      query.where(
        "(LOWER(quiz.title) LIKE :search OR LOWER(user.username) LIKE :search)",
        { search: `%${search.toLowerCase()}%` },
      );
    }

    const total = await query.getCount();

    const items = await query
      .orderBy("quiz.createdAt", "DESC")
      .skip((page - 1) * pageSize)
      .take(pageSize)
      .getMany();

    return { items, total };
  }

  async getTopQuizzesByAttempts(limit: number): Promise<TopQuizDto[]> {
    // Note: This would require joining with QuizAttempts. Adjust according to your attempts entity.
    // This is a placeholder; you'll need to implement based on your actual attempt tables.
    const query = this.quizRepo
      .createQueryBuilder("quiz")
      .leftJoin("quiz.user", "user")
      .leftJoin("quiz.attempts", "attempt") // assuming attempts relation exists
      .select([
        "quiz.id AS id",
        "quiz.title AS title",
        "user.username AS authorUsername",
        "COUNT(attempt.id) AS attemptCount",
        // Add comment count, question count if needed
      ])
      .where("quiz.visibility = :visibility", { visibility: "public" })
      .andWhere("quiz.disabled = false")
      .groupBy("quiz.id, user.username")
      .orderBy("attemptCount", "DESC")
      .limit(limit);

    const results = await query.getRawMany();
    // Map to TopQuizDto
    return results.map((r) => ({
      id: r.id,
      title: r.title,
      authorUsername: r.authorUsername,
      attemptCount: parseInt(r.attemptCount, 10),
      commentCount: 0, // you need to join comments
      questionCount: 0,
    }));
  }

  async removeQuizTags(quizId: number): Promise<void> {
    await this.quizTagRepo.delete({ quizId });
  }
}
